flowmachine/flowmachine/features/subscriber/active_subscribers.py

```diff
@@ -0,0 +1,170 @@
+        of the query
+    period_unit: {'days','hours','minutes'} default 'hours'
+        The unit of time to of minor_period_length
+    subscriber_identifier : {'msisdn', 'imei'}, default 'msisdn'
+        Either msisdn, or imei, the column that identifies the subscriber.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+    hours: tuple default None
+        A range of hours to restrict contributions to minor_period_counts to.
+
+    Notes
+    -----
+    * if minor_period_length is equal to or less than the length of hours, there will be a set
+    of major_periods that are guaranteed to be empty (as they fall outside the range set by hours).
+
+    Examples
+    --------
+    Returns subscribers who were active for at least one hour a day for at least three days
+    between 2016-01-01 and 2016-01-04
+
+    >>>    active_subscribers = ActiveSubscribers(
+                start_date=date(year=2016, month=1, day=1),
+                minor_period_length=1,
+                minor_periods_per_major_period=24,
+                total_major_periods=4,
+                minor_period_threshold=1,
+                major_period_threshold=3,
+                tables=["events.calls"],
+            )
+
+    Returns subscribers that were active in at least two ten minute intervals within half an hour,
+    at least three times across the two hours between 20:00:00 and 22:00:00 on 2016-01-01
+
+    >>>     active_subscribers = ActiveSubscribers(
+                start_date=datetime(year=2016, month=1, day=1, hour=20),
+                minor_period_length=10,
+                minor_periods_per_major_period=3,
+                total_major_periods=4,
+                minor_period_threshold=2,
+                major_period_threshold=3,
+                period_unit="minutes",
+                tables=["events.calls"],
+            )
+
+
+    """
+
+    period_to_rrule_mapping = {
+        "days": rr.DAILY,
+        "hours": rr.HOURLY,
+        "minutes": rr.MINUTELY,
+    }
+
+    def __init__(
+        self,
+        start_date: Union[date, datetime, str],
+        minor_period_length: int,
+        minor_periods_per_major_period: int,
+        total_major_periods: int,
+        minor_period_threshold: int,
+        major_period_threshold: int,
+        period_unit: Literal["days", "hours", "minutes"] = "hours",
+        subscriber_identifier: Optional[str] = "msisdn",
+        tables: Optional[List[str]] = None,
+        subscriber_subset: Optional[SubscriberSubsetType] = None,
+        hours: Optional[Tuple[int, int]] = None,
+    ):
+        self.start_date = start_date
+        self.minor_period_threshold = minor_period_threshold
+        self.sub_id_column = subscriber_identifier
```

Comment:
Up to you on this one, but I just want to note:

Regarding which parameters to set as attributes, the only essential thing is that no two queries with different parameters end up with exactly the same attributes (because then they would both have the same query ID). In this case, `major_period_threshold` is the only parameter that _needs_ to be set as an attribute, because all the others are captured within the `major_period_queries` sub-queries.

There's no problem with setting other parameters as attributes too, and in some cases it may be useful for a user (or possibly another query using this as a sub-query) to check these attributes rather than needing to search through the sub-queries. For example, I think `start_date` is a useful attribute here even though it's not essential (and same for `end_date`, as per my comment above). On the other hand, I personally wouldn't keep the `sub_id_column` attribute. I'm fairly neutral on the various major/minor period parameters (other than `major_period_threshold`) - I'll let you make the call on those.